File: Sliding_Window/LC_424.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.28

class Solution {
public:
    int characterReplacement(string s, int k) {
        // two pointers
        int l = 0, r = 0;
        int ans = 0;
        int max_count = 0;
        unordered_map<char, int> count;
        while (r < s.size()){
            count[s[r]]++;
            max_count = max(max_count, count[s[r]]);
            // if current substring length is greater than max count + k, need to shift the left pointer!
            if ((r-l+1) > max_count + k){
                count[s[l]]--;
                l++;
            }
            ans = max(ans, r-l+1);
            r++;
        }
        return ans;
    }
};

File: Sliding_Window/LC_424.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.28

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # two pointers
        l, r = 0, 0
        count = collections.defaultdict(int)
        ans = 0
        max_f = 0
        while r < len(s):
            count[s[r]] += 1
            max_f = max(max_f, count[s[r]])
            if r-l+1 > max_f + k: # current interval length is greater than max freq + k
                # shift left pointer
                count[s[l]] -= 1
                l += 1
            ans = max(ans, r-l+1)
            r += 1
        return ans